src/core/chart-transformers.js

```diff
@@ -0,0 +1,287 @@
+
+/*
+ * Graph transformation function that accepts two keywords (x and y) and a graph
+ * containing one or more pairs of timeseries and combines pairs of matching time 
+ * series into a variable response graph.
+ * 
+ * Each data series should match exactly one other series. In order to match, two 
+ * data series must:
+ *   - have names that are identical except for the substitution of x for y
+ *   - have data at all the same timestamps
+ * 
+ * This function combines each pair of matching data series into a new data series. For
+ * each (time, data) tuple present in both original time series, it creates a new
+ * (data-x, data-y) tuple, using the series with the x keyword as the x coordinate
+ * and the series with the y keyword as the y coordinate.
+ *
+ * The axis labels of the new graph will be generated from the y axis label(s) of the
+ * old graph.
+ *
+ * Example:
+ * x: pr
+ * y: tasmax
+ * chart with data.columns:
+ * ["Monthly pr", 10, 20, 30, 40, 50 ]
+ * ["Monthly tasmax", 1, 2, 3, 4, 5 ]
+ * ["x", 1/1/15, 1/2/15, 1/3/15, 1/4/15, 1/5/15]
+ * 
+ * Would result in a new chart with data.columns:
+ * ["x", 10, 20, 30, 40, 50]
+ * ["pr", 1, 2, 3, 4, 5]
+ * 
+ * This is intended to graph the effect of one variable (x) on another (y).
+ */
+function makeVariableResponseGraph (x, y, graph) {
+  let c3Data = {};
+
+  const seriesNameContains = function (series, keyword) {
+    return caseInsensitiveStringSearch(series[0], keyword);
+  }
+  
+  const xseries = _.filter(graph.data.columns, series => seriesNameContains(series, x));
+  const yseries = _.filter(graph.data.columns, series => seriesNameContains(series, y));
+    
+  let tuples = [];
+  
+  for(let i = 0; i < xseries.length; i++) {
+    //Try to match each dependent variable series with an independent variable series
+    let dependent = xseries[i];
+    let independent = _.find(yseries, series => {
+      return series[0].toLowerCase().replace(y.toLowerCase(), x.toLowerCase()) === 
+        dependent[0].toLowerCase();
+      });
+    for(let d = 1; d < dependent.length; d++) {
+      if(!_.isNull(dependent[d]) && !_.isNull(independent[d])) {
+        tuples.push([independent[d], dependent[d]]);
+      }
+    }
+  }
+  //sort by x value, preperatory to putting on the graph.
+  tuples.sort((a, b) => a[0] - b[0]);  
+  c3Data.columns = [["x"], [y]];
+
+
+  for(let i = 1; i < tuples.length; i++) {
+    c3Data.columns[0].push(tuples[i][0]);
+    c3Data.columns[1].push(tuples[i][1]);
+    //C3 doesn't really support scatterplots, but we can fake it by adding
+    //a missing data point between each actual data point, and instructing C3
+    //not to connect across missing data points with {connectNull: false} 
+    //TODO: breake this out into makeScatterplot(); we'll likely need it again
+    if(i < tuples.length - 1) {
+      c3Data.columns[0].push((tuples[i][0] + tuples[i+1][0])/2);
+      c3Data.columns[1].push(null);
+    }
+  }
```

Comment:
Wherever you find yourself iterating through an array and pushing items onto another array derived from the iterated array, you can usually reformulate it more clearly as a set of filter, map, and other operations (courtesy of underscore). 

When the set of filter and map operations gets larger that 2 or 3 (or even 1!), you can use chaining (part of underscore) to make it clearer: Chaining means that the array arguments you'd have to pass explicitly to the filter, map, etc. functions are omitted because the operations are methods of the array (wrapped as a chainable object by underscore). It's very nice, and close to the syntax of some of the more delightful functional languages.

Reply:
I see your point; I didn't quite notice the double pushes. Just for reference, the double pushes could be accommodated by mapping to pairs (two-element arrays) and then flattening the result. (In this case, you'd also need to push the final two results afterwards. That plus other minor complications makes my functional solution sound worse than your iterative solution in this case.)

I don't see your reduce solution here, but it doesn't _sound_ like a better solution.

Reply:
Sorry about that; I thought github would update with what I'd changed this section to. [Here's the reduce solution.](https://github.com/pacificclimate/climate-explorer-frontend/pull/146/files#diff-dc15f875105ec95e990b237fac511880R94) I wasn't terribly happy with it.